File: project_euler/Problem038/Problem038.java
package project_euler;

import java.util.Arrays;

public class Problem038 {
    public void solution() {
        boolean[] digits = new boolean[9];
        for (int i = 9487; i >= 9234; --i) {
            Arrays.fill(digits, false);
            if (isPandigital(i, digits) && isPandigital(i << 1, digits)) {
                System.out.printf("%d%d\n", i, i << 1);
                return;
            }
        }
    }

    private boolean isPandigital(int value, boolean[] digits) {
        while (0 != value) {
            int index = value % 10 - 1;
            if (index < 0 || digits[index]) {
                return false;
            }
            digits[index] = true;
            value /= 10;
        }
        return true;
    }
}

File: project_euler/Problem038/Problem038.go
package project_euler

import (
    "fmt"
)

func Problem038() {
    digits := make([]bool, 9)
    for i := 9487; i >= 9234; i-- {
        for j := 0; j < 9; j++ {
            digits[j] = false
        }

        if (isPandigital(i, digits) && isPandigital(i << 1, digits)) {
            fmt.Printf("%d%d\n", i, i << 1)
            return
        }
    }
}

func isPandigital(value int, digits []bool) bool {
    for 0 != value {
        index := value % 10 - 1
        if index < 0 || digits[index] {
            return false
        }
        digits[index] = true
        value /= 10
    }
    return true
}
